src/config/datadog.rs

```diff
@@ -0,0 +1,117 @@
+use super::{ComponentId, Config, SinkOuter, SourceOuter};
+use crate::{
+    sinks::datadog::metrics::DatadogConfig, sources::internal_metrics::InternalMetricsConfig,
+};
+use serde::{Deserialize, Serialize};
+
+static INTERNAL_METRICS_KEY: &str = "#datadog_internal_metrics";
+static DATADOG_METRICS_KEY: &str = "#datadog_metrics";
```

Comment:
Do we expect this to change in https://github.com/timberio/vector/pull/8676 to allow for "internal" components?

Reply:
Currently, `ComponentScope` as implemented in #8676 allows for a `Public` and `Scope` -- there isn't a concept of scope also being 'internal'.

IIRC, there was some discussion about having a separate 'visibility' identifier that could answer this use case.

@jdrouet, is that accurate? Do you anticipate us separately adding visibility outside of the notion of 'scope'?

Reply:
For now the notion of scope is only used for pipelines (I'm introduction `pipeline_id#transform_id` in https://github.com/timberio/vector/pull/8810) but I guess we could had another scope `Internal` that would `to_string` to something like `#component_id`.

The other solution could be to add a separate field in `ComponentId` but I don't know if it would add some value 🤔

Reply:
I was thinking of something akin to segregated scope for 'internal' components.

The idea being, that even if we have both a user-defined and an 'internal' component named "a", they wouldn't clash with each other and cause topology to fail.

I imagine using special symbols internally (such as "#" in this case) won't go far enough, unless we also make those unspeakable in user config.

Reply:
> The idea being, that even if we have both a user-defined and an 'internal' component named "a", they wouldn't clash with each other and cause topology to fail.

This is already a different behavior that the pipeline scope 🤔 If transform `foo` is defined globally and `foo` is also defined in pipeline `bar`, it should fail. But that's checked when compiling `Config` so it can be skipped.

Reply:
I think this is worth addressing.

The `#` character acting as an separator / internal designator won't save us from user-configured components of the same name. It's a happy accident that it's not a valid TOML key name, but as you mentioned, it's still valid JSON.

Ideally, we can wind up with a scenario where anything 'internally' created -- scoped components, internally managed pipelines - cannot clash with user config, even if the labels we use happen to overlap.

This probably means changing some of the validation logic so that name tests are solely against the `Public` scope and not other enum arm.